.circleci/config.yml

```diff
@@ -0,0 +1,64 @@
+version: 2.1
+
+jobs:
+  build:
+    docker:
+      # see https://hub.docker.com/r/codeaches/openjdk
+      - image: codeaches/openjdk:12-jdk
+
+    working_directory: ~/repo
+
+    environment:
+      JVM_OPTS: -Xmx3200m
+      TERM: dumb
+
+    steps:
+      - checkout
+
+      - restore_cache:
+          keys:
+            - v1-dependencies-{{ checksum "build.gradle" }}
+            - v1-dependencies-
+
+      - run: gradle dependencies
+
+      - save_cache:
+          paths:
+            - ~/.gradle
+          key: v1-dependencies-{{ checksum "build.gradle" }}
+
+      - run: gradle build
+
+      - persist_to_workspace:
+          root: ~/repo/jar
+          paths:
+            - httpserver-1.0-SNAPSHOT.jar
+
+
+  deploy:
+    machine: true
+    steps:
+
+      - attach_workspace:
+          at: /tmp/workspace
+
+      - add_ssh_keys:
+          fingerprints:
+            - "78:11:cd:e5:32:0f:71:41:27:c9:a9:2a:5e:3c:d9:91"
+
+      - run:
+          name: Send JAR to Server
+          command: |
+            scp /tmp/workspace/httpserver-1.0-SNAPSHOT.jar $SSH_USER@$SSH_HOST:$SSH_DEPLOY_PATH
```

Comment:
I actually don't know a ton about how Java handles this, but I would be surprised if it's sufficient to just overwrite the jar on the remote system and keep running—I would expect that once the JVM classloader has read data from the old jar, it won't re-load it without a restart or some kind of signal. Have you been able to verify that new code actually runs at the end of this pipeline? I'd search for something like "hot swapping jar" or "hot swapping JVM" to figure this out.

Reply:
Two other thoughts:

1) It would be cool to work out a way to deploy new code without any disruption to your app. There are lots of strategies for this, usually involving deploying your new build artifact (the new jar) while the old one keeps running, and then updating a symlink or similar to point to the latest version. This has advantages for rolling back, too—if you have the old jar and the new one deployed right next to each other, you can roll back to the previous version easily if something goes wrong. I'd search for "zero downtime deployment" and "zero downtime deployment Java jar" for more resources on this concept.

2) One thing that saves my butt all the time and helps me answer "is the newest code actually running" is an endpoint in my deployed app that returns some unique identifier, like the git rev of the currently deployed code, or a version number that I manually bump in the source code. This lets you really easily answer "what code sent me this response" without having to look on the server. I often put this information in a general [health check](https://microservices.io/patterns/observability/health-check-api.html) endpoint that other tools like a load balancer or monitoring tool might consume, something like `/_healthcheck/` or similar.